File: src/frontend/src/features/manual-qc/types/snapshot.ts
export interface SnapshotQuery {
  stat_date_start?: string
  stat_date_end?: string
  scene_name?: string
  group_name?: string
  employee_id?: string
  project_name?: string
}

export interface MetricTotals {
  annotation_total: number
  annotation_submitted: number
  expect_alloc: number
  actual_alloc: number
  actual_complete: number
  correct: number
  incorrect: number
  expect_pass: number
  expect_reject: number
  actual_pass: number
  actual_reject: number
}

export interface MetricState extends MetricTotals {
  conclusion: 'pass' | 'reject' | 'pending' | null
  exec_status:
    | 'PENDING'
    | 'PASS_EXECUTING'
    | 'PASS_DONE'
    | 'REJECT_EXECUTING'
    | 'REJECT_DONE'
    | null
}

export interface ProjectAggregate {
  stat_date: string
  project_name: string
  annotation_total: number
  annotation_submitted: number
  good_metrics: MetricTotals
  bad_metrics: MetricTotals
  computed_at: string
}

export interface SceneAggregate extends ProjectAggregate {
  scene_name: string
}

export interface GroupAggregate extends SceneAggregate {
  group_name: string
}

export interface EmployeeAggregate extends GroupAggregate {
  employee_id: string
}

export interface SnapshotRow {
  id: number
  stat_date: string
  scene_name: string
  group_name: string
  employee_id: string
  project_name: string
  annotation_total: number
  annotation_submitted: number
  good_metrics: MetricState
  bad_metrics: MetricState
  option_metrics: Record<string, Record<string, MetricState>>
  confirmed_by: string | null
  confirmed_at: string | null
  executed_by: string | null
  executed_at: string | null
  execution_note: string | null
  computed_at: string
  updated_at: string
}

export interface DataResponse<T> {
  schema_version: 'snapshot-jsonb-v20260709'
  items: T[]
  total: number
  computed_at: string | null
}

export type AggregateLevel = 'project' | 'scene' | 'group' | 'employee'

export interface AggregateNode extends ProjectAggregate {
  id: string
  level: AggregateLevel
  scene_name: string
  group_name: string
  employee_id: string
  children?: AggregateNode[]
  hasChildren: boolean
  loading?: boolean
}

File: src/api/schemas/snapshot.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Generic, Literal, TypeVar

from pydantic import BaseModel, ConfigDict, Field


class HealthResponse(BaseModel):
    status: str
    app: str
    schema_version: str
    database_alias: str
    database: str
    postgres_version: str


class MetricTotals(BaseModel):
    annotation_total: int = Field(ge=0)
    annotation_submitted: int = Field(ge=0)
    expect_alloc: int = Field(ge=0)
    actual_alloc: int = Field(ge=0)
    actual_complete: int = Field(ge=0)
    correct: int = Field(ge=0)
    incorrect: int = Field(ge=0)
    expect_pass: int = Field(ge=0)
    expect_reject: int = Field(ge=0)
    actual_pass: int = Field(ge=0)
    actual_reject: int = Field(ge=0)


class MetricState(MetricTotals):
    conclusion: Literal["pass", "reject", "pending"] | None = None
    exec_status: Literal[
        "PENDING",
        "PASS_EXECUTING",
        "PASS_DONE",
        "REJECT_EXECUTING",
        "REJECT_DONE",
    ] | None = None


class SceneAggregateOut(BaseModel):
    stat_date: date
    scene_name: str
    annotation_total: int = Field(ge=0)
    annotation_submitted: int = Field(ge=0)
    good_metrics: MetricTotals
    bad_metrics: MetricTotals
    computed_at: datetime


class GroupAggregateOut(SceneAggregateOut):
    group_name: str


class EmployeeAggregateOut(GroupAggregateOut):
    employee_id: str


class SnapshotRowOut(BaseModel):
    """The current 18-column snapshot contract."""

    model_config = ConfigDict(extra="forbid")

    id: int
    stat_date: date
    scene_name: str
    group_name: str
    employee_id: str
    project_name: str
    annotation_total: int = Field(ge=0)
    annotation_submitted: int = Field(ge=0)
    good_metrics: MetricState
    bad_metrics: MetricState
    option_metrics: dict[str, dict[str, MetricState]]
    confirmed_by: str | None = None
    confirmed_at: datetime | None = None
    executed_by: str | None = None
    executed_at: datetime | None = None
    execution_note: str | None = None
    computed_at: datetime
    updated_at: datetime


T = TypeVar("T")


class DataResponse(BaseModel, Generic[T]):
    schema_version: str = "snapshot-jsonb-v20260709"
    items: list[T]
    total: int = Field(ge=0)
    computed_at: datetime | None = None
